code/FormSchemaController.php

```diff
@@ -0,0 +1,162 @@
+<?php
+
+namespace SilverStripe\Admin;
+
+use SilverStripe\Control\HTTPRequest;
+use SilverStripe\Control\HTTPResponse;
+use SilverStripe\Core\Validation\ValidationResult;
+use SilverStripe\Forms\Form;
+use SilverStripe\Forms\Schema\FormSchema;
+
+/**
+ * The base class for controllers in the CMS that need access to form schema functionality.
+ */
+abstract class FormSchemaController extends AdminController
+{
+    /**
+     * Form schema header identifier
+     */
+    public const SCHEMA_HEADER = 'X-Formschema-Request';
+
+    private static array $allowed_actions = [
+        'schema',
+        'methodSchema',
+    ];
+
+    private static array $url_handlers = [
+        'GET schema/$FormName/$ItemID/$OtherItemID' => 'schema',
+        'GET methodSchema/$Method/$FormName/$ItemID' => 'methodSchema',
+    ];
+
+    private static array $dependencies = [
+        'FormSchema' => '%$' . FormSchema::class,
+    ];
+
+    /**
+     * Current form schema helper
+     */
+    private ?FormSchema $schema = null;
+
+    /**
+     * Get form schema helper
+     */
+    public function getFormSchema(): FormSchema
+    {
+        if (!$this->schema) {
+            $this->schema = FormSchema::singleton();
+        }
+        return $this->schema;
+    }
+
+    /**
+     * Set form schema helper for this controller
+     */
+    public function setFormSchema(FormSchema $schema): static
+    {
+        $this->schema = $schema;
+        return $this;
+    }
+
+    /**
+     * Gets a JSON schema representing the current edit form.
+     */
+    public function schema(HTTPRequest $request): HTTPResponse
+    {
+        $formName = $request->param('FormName');
+        $itemID = $request->param('ItemID');
+
+        if (!$formName) {
+            $this->jsonError(400, 'Missing request params');
+        }
+
+        // The getter can accept an ID where the main form action wouldn't
+        $formMethod = "get{$formName}";
+        if (!$this->hasMethod($formMethod)) {
+            $formMethod = $formName;
+            if (!$this->hasMethod($formMethod)) {
+                $this->jsonError(404, 'Form not found');
+            }
+        }
```

Comment:
Most usages of this method right now (e.g. see AssetAdmin) have two methods - one method is named the same as the form, and does not accept a parameter - the other prefixed with `get` and accepts a parameter.

The usages of `ModalController` used to go through `methodSchema` on `LeftAndMain` - but now we just go directly through `schema` on `ModalController`. Those methods do _not_ have `get` prefixes.

I'm not going through and refactoring all of those - that's out of scope. Instead, I've just updated this method to account for both scenarios.

Reply:
Could you put some the github comment info into the inline comment - by itself the existing comment "The getter can accept an ID where the main form action wouldn't" is pretty confusing

Reply:
Happy to do so, but I'm not sure what you find confusing about it so I don't know what info you want in there. What would you like the comment to say? Or what about the existing comment is confusing?

Reply:
Most usages of this method (e.g. AssetAdmin) have two methods - one method is named the same as the form, and does not accept a parameter - the other prefixed with "get" and accepts an `$itemID` parameter.

^ Maybe just the comment to this